pip-lock/main.ts

```diff
@@ -0,0 +1,392 @@
+
+            if(current.length < 4)
+                throw new Error(`invalid input - requirement list too short ${JSON.stringify(current)}`);
+
+            const packageNameRaw = current[2];
+            const packageVersionRaw = current[current.length - 1];
+
+            const nameMatches = packageNameRaw.match(/([A-Za-z0-9\.\-_]+)(.*)/);
+            if(!nameMatches || !nameMatches.length)
+                throw new Error(`invalid package name gathered as locally satisfied requirement: ${packageNameRaw}`);
+
+            const name = nameMatches[1];
+            const version = packageVersionRaw.replace('(', '').replace(')', '');
+            res.push({name: name, version: version});
+
+            return {
+                state: 'requirement',
+                name: null,
+                version: null,
+            };
+        },
+        'Would': (current: string[], last: ParseState): ParseState => {
+            // Not currently used, but included here for future use; this would receive a list
+            // of _new_ packages only that would be installed; we are currently just capturing the total
+            // list. Example input: `Would install Django-4.1 asgiref-3.5.2 backports.zoneinfo-0.2.1 sqlparse-0.4.2`
+            // Which means current would contain:
+            // `["install", "Django-4.1", "asgiref-3.5.2", "backports.zoneinfo-0.2.1", "sqlparse-0.4.2"]`
+            return {
+                state: 'end',
+                name: null,
+                version: null,
+            };
+        },
+    };
+
+    // This loop walks the lines of output from the pip dry run.
+    // We process each line we have a defined action for (in the ParseMap above),
+    // and skip ones that we don't. This could include console output with download
+    // progress indicators, and other similar artifacts that we will essentially drop.
+    let lastOutput: ParseState = {state: 'initial', name: null, version: null};
+    for(let i = 0; i < unpacked.length; i++) {
+        const trimmed = unpacked[i].trim();
+        const current: string[] = trimmed.split(" ");
+        if(!current.length)
+            throw new Error("invalid input provided");
+
+        const key = current[0];
+        const action: NullableParseStep = actions[key];
+
+        if(!action)
+            continue;
+
+        lastOutput = action(current.slice(1), lastOutput);
+    }
+
+
+    return res;
+}
+
+// We will essentially run pip with the provided arguments, parse the package output, and return the package list.
+async function checkAndRunPip(args, pipPath: string): Promise<PackageList> {
+    let packages: PackageEntry[] = [];
+
+    const pipVersion = await getVersion(pipPath);
+    if(pipVersion.major < 22 || pipVersion.minor < 2 || pipVersion.fix < 2) {
+        throw new Error(
+            `pip version ${pipVersion.major}.${pipVersion.minor}.${pipVersion.fix} found is too old, please upgrade to at least 22.2.2`);
+    }
```

Comment:
Based on this conditional, you would reject `23.0.0` because it satisfies `pipVersion.minor < 2`